crates/pico_core/src/dyn_eq.rs

```diff
@@ -0,0 +1,45 @@
+use std::any::Any;
+
+pub trait DynEq: Any {
+    fn as_any(&self) -> &dyn Any;
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+    fn dyn_eq(&self, other: &dyn DynEq) -> bool;
+}
+
+impl<T> DynEq for T
+where
+    T: Any + Eq,
+{
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    fn dyn_eq(&self, other: &dyn DynEq) -> bool {
+        if let Some(other) = other.as_any().downcast_ref::<T>() {
+            return self == other;
+        }
+        false
+    }
+}
+
+impl PartialEq<dyn DynEq> for dyn DynEq {
+    fn eq(&self, other: &dyn DynEq) -> bool {
+        self.dyn_eq(other)
+    }
+}
+
+impl PartialEq<&Self> for Box<dyn DynEq> {
+    fn eq(&self, other: &&Self) -> bool {
+        <Self as PartialEq>::eq(self, *other)
+    }
+}
+
+impl core::fmt::Debug for dyn DynEq {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "DynEq {{ .. }}")
```

Comment:
This might sound annoying, but can we require Any + Debug? If we don't want that in general, can we require it in `cfg[test]`, and have this fallback debug impl in non-test?

It's really hard to know what's going on from 
```
            NodeId {
                key: Key(
                    17133476938307633808,
                ),
                param_id: ParamId(
                    6585809115041605177,
                ),
            }: DynEq { .. },
```

Reply:
this is surprisingly hard